Entity Query access control to decide if users can reference content.

ekes opened the pull request:
https://github.com/localgovdrupal/localgov_services/issues/216#issuecomment-3223678278

Adds tests. I think they're failing, because I think they do get the content in the list.

Pushing for work done. @andybroomfield

ekes committed:
Entity Query access control to decide if users can reference content.

https://github.com/localgovdrupal/localgov_services/issues/216#issuecomment-3223678278

Adds tests. I think they're failing, because I think they do get the
content in the list.

ekes committed:
Unused use statement.

andybroomfield commented:
This is what I am seeing with this patch.

Testing with an unpublished landing page
<img width="920" height="216" alt="Screenshot 2025-08-27 at 3 41 38 pm" src="https://github.com/user-attachments/assets/cd7e45c9-c2dd-4f02-a289-8f484ee0ec42" />

Before: Editor cannot reference unpublished landing pages.
<img width="803" height="319" alt="Screenshot 2025-08-27 at 3 41 46 pm" src="https://github.com/user-attachments/assets/ef65543b-e7fd-4848-af99-461d6a00664d" />
(Nothing in the entity reference drop down and this can't be saved).

After: Unpublished landing page appears in drop down autocomplete. 
<img width="810" height="425" alt="Screenshot 2025-08-27 at 3 49 19 pm" src="https://github.com/user-attachments/assets/f8b04c12-30de-4155-ba42-bdc45aedd8f2" />

And I can link on the landing page.
<img width="846" height="491" alt="Screenshot 2025-08-27 at 3 49 59 pm" src="https://github.com/user-attachments/assets/80301e9f-f5b4-41d4-affa-dc4109149c93" />

ekes commented:
What the tests are at least failing on is that users that _don't_ have the permission to see the unpublished content _also_ see it.

andybroomfield commented:
@ekes ok, looking at that. I took away the editors view any unpublished content and checked again, whilst they can't see the unpublished landing page, it still appears in the list.
Is the check being run here for `view` or `view_label` as I know that was a recent Drupal change.

andybroomfield commented:
I think the patch we may have added [Users without 'bypass node access' permission can't reference unpublished content even if they have access to it](https://www.drupal.org/project/drupal/issues/2845144#comment-15554276) might be causing a conflict with this. 

Maybe the safer option is to leave the existing code but add a check for hasPermission('view any unpublished content') alongside the checks for node acesss?

andybroomfield commented:
Interesting that whilst I could save it, going back to the edit page gave me this.... (Parent is now restricted access).
<img width="921" height="664" alt="Screenshot 2025-08-27 at 6 35 12 pm" src="https://github.com/user-attachments/assets/8c2cad82-2efc-444c-839e-aa992b1d5a51" />

andybroomfield commented:
Just noting that a [similar fix](https://www.drupal.org/project/drupal/issues/2807629) was [committed](https://git.drupalcode.org/project/drupal/-/commit/08177ee3651abc4f83e8619707d70c2fddc5af1e#b951feca59c32347c2efab78b8ccf424644b807a_152_167) for menu items in Drupal 10.1 so whilst I don't know why `entityAccess(TRUE)` is not filtering, adding `!$this->account->hasPermission('view any unpublished content')` to the check looks like the correct way forward.

andybroomfield committed:
Restore the previous node_access check with addition of view any unpublished permission

In the entity reference service selector.
Restore previous node_access checking, and restrict to published nodes if the checks
do not pass. Additionally, check if the user has permission view any unpublished content
and allow access to unpublished nodes if that is the case.

andybroomfield committed:
Apply view any unpublished content permission fix to EntityChildRelationshipsUi

Apply same limit to the widget used to find referenced child pages.

andybroomfield commented:
@ekes I've tried to restore the previous logic and add the permission check for 'view any unpublished content' but that seems to have made the tests fail. From manually testing though it appears to be functional, with unpublished landing / service pages hidden if the user does not have any of the right access permissions.
